File: src/agent/loop.ts
import { compact } from "./compaction";
import { AGENT_CONFIG } from "./config";
import { NoteStore, NoteStoreError } from "./noteStore";
import { buildSystemPrompt } from "./systemPrompt";
import { estimateTokens } from "./tokens";
import {
  executePatchNote,
  executeReadNote,
  executeRewriteNote,
  PATCH_NOTE_SCHEMA,
  READ_NOTE_SCHEMA,
  REWRITE_NOTE_SCHEMA,
} from "./tools";
import { CanonicalMessage, Provider, ProviderError, ToolCall, ToolSchema } from "./types";

export class AgentProviderError extends Error {
  readonly providerError: ProviderError;
  constructor(providerError: ProviderError) {
    super(providerError.message);
    this.name = "AgentProviderError";
    this.providerError = providerError;
  }
}

export type StoppedReason = "completed" | "max_iterations" | "cancelled" | "timed_out";

/**
 * Pass this as the abort reason (`controller.abort(TURN_TIMEOUT)`) when a turn
 * is being cut off by a deadline rather than by the user. Both arrive as an
 * `AbortError`, and reporting a 2-minute timeout as "Cancelled" tells the user
 * they did something they didn't.
 */
export const TURN_TIMEOUT = "anotai:turn-timeout";

export type TurnResult = {
  finalText: string;
  bodyChanged: boolean;
  newBody: string;
  stoppedReason: StoppedReason;
  toolCallsExecuted: number;
  updatedHistory: CanonicalMessage[];
};

export type RunTurnParams = {
  prompt: string;
  noteTitle: string;
  store: NoteStore;
  /** Prior messages for this note session (excludes the system prompt, which is rebuilt every request). Empty for a fresh session. */
  history: CanonicalMessage[];
  provider: Provider;
  signal?: AbortSignal;
  /**
   * Called after each tool call that actually changed the note, with the body
   * either side of that single write. Lets the UI highlight and scroll to each
   * change as it lands rather than only once at the end of the turn (PRD §7.5).
   *
   * A plain callback, not an event emitter: the core stays framework-free, and
   * the caller decides what a change means. Never called for a tool that left
   * the note untouched.
   */
  onNoteWritten?: (before: string, after: string) => void;
};

/** The tools that can change the note, so the others are never read back around. */
const WRITE_TOOLS = new Set(["rewrite_note", "patch_note"]);

function isAbortError(err: unknown): boolean {
  return err instanceof Error && err.name === "AbortError";
}

function selectTools(noteTokenCount: number): ToolSchema[] {
  const tools = [READ_NOTE_SCHEMA, REWRITE_NOTE_SCHEMA];
  if (noteTokenCount >= AGENT_CONFIG.FORCE_REWRITE_BELOW_TOKENS) {
    tools.push(PATCH_NOTE_SCHEMA);
  }
  return tools;
}

/**
 * Dispatches one tool call. Arguments are untrusted model output, so each
 * executor validates its own input and returns a tool error rather than
 * throwing — a bad argument becomes something the model can correct on the next
 * iteration instead of aborting the user's turn.
 *
 * Storage failures and cancellation are the two exceptions: both rethrow, since
 * neither is something the model can fix by trying again.
 */
function executeTool(call: ToolCall, store: NoteStore): unknown {
  try {
    switch (call.name) {
      case "read_note":
        return executeReadNote(store, call.arguments);
      case "rewrite_note":
        return executeRewriteNote(store, call.arguments);
      case "patch_note":
        return executePatchNote(store, call.arguments);
      default:
        return { ok: false, error: `Unknown tool: ${call.name}` };
    }
  } catch (err) {
    if (err instanceof NoteStoreError || isAbortError(err)) throw err;
    return { ok: false, error: err instanceof Error ? err.message : "The tool failed unexpectedly." };
  }
}

/**
 * Runs one agent turn to completion: builds the system prompt from live note
 * state each iteration, streams from the provider, executes any tool calls
 * locally, and repeats until the model replies with plain text or a guard
 * trips (PRD §6.6). The editor is expected to be soft-locked by the caller
 * for the duration of this call.
 */
export async function runTurn(params: RunTurnParams): Promise<TurnResult> {
  const { prompt, noteTitle, store, provider } = params;
  const signal = params.signal ?? new AbortController().signal;

  let messages: CanonicalMessage[] = [...params.history, { role: "user", content: prompt }];
  const snapshot = store.read();

  let finalText = "";
  let stoppedReason: StoppedReason = "completed";
  let toolCallsExecuted = 0;

  for (let iteration = 0; iteration < AGENT_CONFIG.MAX_ITERATIONS; iteration++) {
    if (estimateTokens(JSON.stringify(messages)) > AGENT_CONFIG.COMPACT_THRESHOLD_TOKENS) {
      messages = await compact(messages, provider);
    }

    const currentBody = store.read();
    const system = buildSystemPrompt({ title: noteTitle, body: currentBody });
    const tools = selectTools(estimateTokens(currentBody));

    let assistantText = "";
    const toolCalls: ToolCall[] = [];
    let sawError: ProviderError | null = null;

    try {
      for await (const event of provider.send({ system, messages, tools, signal })) {
        if (event.type === "text") assistantText += event.delta;
        else if (event.type === "toolCall") toolCalls.push(event.call);
        else if (event.type === "error") sawError = event.error;
      }
    } catch (err) {
      if (isAbortError(err)) return abortResult();
      throw err;
    }

    // An abort surfaces either as a thrown AbortError (mid-stream) or as an
    // error event (if it lands while the request is still being opened). Both
    // are the same user-visible situation, so they must report identically —
    // checking the signal here is what keeps them from diverging.
    if (sawError) {
      if (signal.aborted) return abortResult();
      throw new AgentProviderError(sawError);
    }

    if (toolCalls.length === 0) {
      messages.push({ role: "assistant", content: assistantText });
      finalText = assistantText;
      stoppedReason = "completed";
      return buildResult();
    }

    messages.push({ role: "assistant", toolCalls });
    for (const call of toolCalls) {
      // Only read back around tools that can write, and only when someone is
      // listening — read_note can't change anything, so comparing before and
      // after it would be two wasted filesystem reads per call.
      const watching = params.onNoteWritten !== undefined && WRITE_TOOLS.has(call.name);
      const before = watching ? store.read() : null;

      const result = executeTool(call, store);
      toolCallsExecuted++;

      if (before !== null) {
        const after = store.read();
        if (after !== before) params.onNoteWritten?.(before, after);
      }

      messages.push({ role: "tool", toolCallId: call.id, content: JSON.stringify(result) });
    }

    if (iteration === AGENT_CONFIG.MAX_ITERATIONS - 1) {
      stoppedReason = "max_iterations";
      finalText = "Stopped after 8 steps.";
    }
  }

  return buildResult();

  /** Shared exit for both abort paths, so a timeout is never reported as a cancellation. */
  function abortResult(): TurnResult {
    const timedOut = signal.reason === TURN_TIMEOUT;
    stoppedReason = timedOut ? "timed_out" : "cancelled";
    finalText = timedOut ? "Timed out" : "Cancelled";
    // An aborted turn is discarded, not recorded. The caller reverts the note to
    // its pre-turn body (see `useAgentTurn`), so keeping the prompt and whatever
    // partial output arrived would leave the history describing edits that no
    // longer exist — and the next turn would ask the model to build on them.
    //
    // It also matters to the wire format: a turn aborted before the model
    // replied leaves history ending on a `user` message, so the next turn
    // appends a second one. OpenAI-shaped APIs tolerate that; Anthropic rejects
    // consecutive same-role messages outright.
    return { ...buildResult(), updatedHistory: params.history };
  }

  function buildResult(): TurnResult {
    const newBody = store.read();
    return {
      finalText: finalText || "Done",
      bodyChanged: newBody !== snapshot,
      newBody,
      stoppedReason,
      toolCallsExecuted,
      updatedHistory: messages,
    };
  }
}

File: lib/features/ai/data/agent/loop.dart
import 'dart:async';
import 'dart:convert';

import 'compaction.dart';
import 'config.dart';
import 'note_store.dart';
import 'system_prompt.dart';
import 'tokens.dart';
import 'tools.dart';
import 'types.dart';

/// A provider failure that should surface to the UI (auth, rate limit, model
/// missing, tool support rejection, …) rather than be treated as a bug.
class AgentProviderError implements Exception {
  const AgentProviderError(this.providerError);
  final ProviderError providerError;

  @override
  String toString() => providerError.message;
}

enum StoppedReason { completed, maxIterations, cancelled, timedOut }

/// Pass this as the cancellation token reason when a turn is being cut off by a
/// deadline rather than by the user. Both arrive as a cancellation, and
/// reporting a 2-minute timeout as "Cancelled" would tell the user they did
/// something they didn't.
const turnTimeoutReason = 'anotai:turn-timeout';

class TurnResult {
  const TurnResult({
    required this.finalText,
    required this.bodyChanged,
    required this.newBody,
    required this.stoppedReason,
    required this.toolCallsExecuted,
    required this.updatedHistory,
  });

  final String finalText;
  final bool bodyChanged;
  final String newBody;
  final StoppedReason stoppedReason;
  final int toolCallsExecuted;
  final List<CanonicalMessage> updatedHistory;
}

class RunTurnParams {
  const RunTurnParams({
    required this.prompt,
    required this.noteTitle,
    required this.store,
    required this.history,
    required this.provider,
    this.cancellation,
    this.onNoteWritten,
  });

  final String prompt;
  final String noteTitle;
  final NoteStore store;

  /// Prior messages for this note session (the system prompt is rebuilt every
  /// request). Empty for a fresh session.
  final List<CanonicalMessage> history;

  final Provider provider;
  final CancellationToken? cancellation;

  /// Called after each tool call that actually changed the note, with the body
  /// either side of that single write, so the UI can highlight and scroll to
  /// each change as it lands. Never called for a tool that left the note
  /// untouched.
  final void Function(String before, String after)? onNoteWritten;
}

List<ToolSchema> _selectTools(int noteTokens) {
  final tools = [readNoteSchema, rewriteNoteSchema];
  if (noteTokens >= AgentConfig.forceRewriteBelowTokens) {
    tools.add(patchNoteSchema);
  }
  return tools;
}

/// Dispatches one tool call. Arguments are untrusted model output, so each
/// executor validates its own input and returns a tool error rather than
/// throwing — a bad argument becomes something the model can correct on the
/// next iteration instead of aborting the user's turn. Storage failures and
/// cancellation are the two exceptions, rethrown since neither is something
/// the model can fix by trying again.
Object _executeTool(ToolCall call, NoteStore store) {
  try {
    switch (call.name) {
      case 'read_note':
        return executeReadNote(store, call.arguments);
      case 'rewrite_note':
        return executeRewriteNote(store, call.arguments);
      case 'patch_note':
        return executePatchNote(store, call.arguments);
      default:
        return {'ok': false, 'error': 'Unknown tool: ${call.name}'};
    }
  } on NoteStoreError {
    rethrow;
  } on CancelledException {
    rethrow;
  } catch (err) {
    return {'ok': false, 'error': err.toString()};
  }
}

String _dumpMessages(List<CanonicalMessage> messages) {
  final buffer = StringBuffer();
  for (final message in messages) {
    switch (message) {
      case UserMessage(:final content):
        buffer.write(content);
      case AssistantMessage(:final content):
        buffer.write(content);
      case AssistantToolCalls(:final toolCalls):
        buffer.write(jsonEncode([
          for (final call in toolCalls)
            {'name': call.name, 'arguments': call.arguments},
        ]));
      case ToolResultMessage(:final content):
        buffer.write(content);
    }
    buffer.write('\n');
  }
  return buffer.toString();
}

/// Runs one agent turn to completion: builds the system prompt from live note
/// state each iteration, streams from the provider, executes any tool calls
/// locally, and repeats until the model replies with plain text or a guard
/// trips. The editor is expected to be soft-locked by the caller for the
/// duration of this call.
Future<TurnResult> runTurn(RunTurnParams params) async {
  final signal = params.cancellation ?? CancellationToken.none();
  var messages = <CanonicalMessage>[
    ...params.history,
    UserMessage(params.prompt)
  ];
  final snapshot = params.store.read();

  var finalText = '';
  var stoppedReason = StoppedReason.completed;
  var toolCallsExecuted = 0;

  // Shared exits, declared before the loop so both call sites stay in reach.
  TurnResult buildResult({List<CanonicalMessage>? updatedHistory}) {
    final newBody = params.store.read();
    return TurnResult(
      finalText: finalText.isEmpty ? 'Done' : finalText,
      bodyChanged: newBody != snapshot,
      newBody: newBody,
      stoppedReason: stoppedReason,
      toolCallsExecuted: toolCallsExecuted,
      updatedHistory: updatedHistory ?? messages,
    );
  }

  TurnResult abortResult() {
    final timedOut = signal.reason == turnTimeoutReason;
    stoppedReason = timedOut ? StoppedReason.timedOut : StoppedReason.cancelled;
    finalText = timedOut ? 'Timed out' : 'Cancelled';
    // An aborted turn is discarded, not recorded. The caller reverts the note
    // to its pre-turn body, so keeping the prompt and whatever partial output
    // arrived would leave history describing edits that no longer exist — and
    // the next turn would ask the model to build on them. It also matters to
    // the wire format: a turn aborted before the model replied leaves history
    // ending on a `user` message, so the next turn appends a second one, which
    // Anthropic rejects outright.
    return buildResult(updatedHistory: params.history);
  }

  for (var iteration = 0; iteration < AgentConfig.maxIterations; iteration++) {
    if (estimateTokens(_dumpMessages(messages)) >
        AgentConfig.compactThresholdTokens) {
      messages = await compact(messages, params.provider);
    }

    final currentBody = params.store.read();
    final system =
        buildSystemPrompt(title: params.noteTitle, body: currentBody);
    final tools = _selectTools(estimateTokens(currentBody));

    var assistantText = '';
    final toolCalls = <ToolCall>[];
    ProviderError? sawError;

    try {
      await for (final event in params.provider.send(
        ProviderRequest(
          system: system,
          messages: messages,
          tools: tools,
          cancellation: signal,
        ),
      )) {
        switch (event) {
          case TextDelta(:final delta):
            assistantText += delta;
          case ToolCallEvent(:final call):
            toolCalls.add(call);
          case ProviderErrorMessage(:final error):
            sawError = error;
          case UsageEvent():
            break;
        }
      }
    } on CancelledException {
      return abortResult();
    }

    if (sawError != null) {
      if (signal.isCancelled) return abortResult();
      throw AgentProviderError(sawError);
    }

    if (toolCalls.isEmpty) {
      messages.add(AssistantMessage(assistantText));
      finalText = assistantText;
      stoppedReason = StoppedReason.completed;
      return buildResult();
    }

    messages.add(AssistantToolCalls(toolCalls));
    for (final call in toolCalls) {
      // Only read back around tools that can write, and only when someone is
      // listening — read_note can't change anything.
      final watching =
          params.onNoteWritten != null && writeToolNames.contains(call.name);
      final before = watching ? params.store.read() : null;

      final result = _executeTool(call, params.store);
      toolCallsExecuted++;

      if (before != null) {
        final after = params.store.read();
        if (after != before) params.onNoteWritten!(before, after);
      }

      messages.add(
          ToolResultMessage(toolCallId: call.id, content: jsonEncode(result)));
    }

    if (iteration == AgentConfig.maxIterations - 1) {
      stoppedReason = StoppedReason.maxIterations;
      finalText = 'Stopped after ${AgentConfig.maxIterations} steps.';
    }
  }

  return buildResult();
}
